nav2_behavior_tree/plugins/condition/goal_updated_condition.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2020 Aitor Miguel Blanco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__GOAL_UPDATED_CONDITION_HPP_
+#define NAV2_BEHAVIOR_TREE__GOAL_UPDATED_CONDITION_HPP_
+
+#include <string>
+
+#include "behaviortree_cpp_v3/condition_node.h"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+
+namespace nav2_behavior_tree
+{
+
+class GoalUpdatedCondition : public BT::ConditionNode
+{
+public:
+  GoalUpdatedCondition(
+    const std::string & condition_name,
+    const BT::NodeConfiguration & conf)
+  : BT::ConditionNode(condition_name, conf)
+  {
+  }
+
+  GoalUpdatedCondition() = delete;
+
+  BT::NodeStatus tick() override
+  {
+    if (status() == BT::NodeStatus::IDLE) {
```

Comment:
Correct me if I'm wrong, but I'm not seeing this kind of logic in the other condition nodes. They all just take in `tick` and return SUCCESS or FAILURE with no checks on status or IDLE. Can you point to docs that you used to create this logic?

Reply:
The only particular feature of condition nodes is that when halted, the node is set to IDLE, and that the halt function cannot be overriten in custom nodes.
Then, all nodes have a status, so we know that any first time the node is ticked (either the first tick ever or after a halt) the status will be IDLE.
I'm not sure though if this is specified in the bt.cpp webpage or readmes. I can check if the information is missing and ask Davide how to update if necessary.

Reply:
If this `tick` is called though, its been set to do something or asking for a response. I don't think a check on idle is needed. If you look at the other control nodes in this repo, none of them do any of this logic.

That's not to say that this is wrong and those are right, but it does mean that if this is the "right" way to do  things, all the other ones need to be updated (not in this PR mind you).

Are you sure this is necessary?

Reply:
Hmm, I don't think the existing nodes need this, but the ones that @naiveHobo is working on do.

For goal reached, we have just the distance to the goal and tolerance. None of these change during execution (okay, the goal does, but it's set in the blackboard).
For initial pose recieved, it's similar, either the initial pose is available or not.
Same with transform available.
In the case of is stuck it's a bit different, because it's measuring the movement of the robot, and it would be possible to add an initialization when it's idle, but since on each iteration the pose updates, it's quite redundant.

In the cases of timeout or distance travelled, we are measuring the distance or time between the initialization of the node and the current state, so it would be necessary to add them there.

To sum up, this check is necessary here and in all nodes that want to measure changes between the state of the first tick on the node (first tick ever or after a halt) and the current situation.
If the condition is independent of the state of the bt in the first tick of the node, then there this is redundant, and unnecessary.

Reply:
Yup, this is the way to go. This check ensures each condition is reset if another condition at the same level returns success. Good catch!